tests/test_serialization.py

```diff
@@ -148,6 +151,35 @@ def test_pickle(obj: Any, ref_type: Any) -> None:
             assert c1._metadata.key_type is Any
 
 
+@pytest.mark.parametrize(  # type: ignore
+    "obj,ref_type,key",
+    [
+        (OmegaConf.structured(DictUnion), Union[int, float], "dict"),
+        (OmegaConf.structured(ListUnion), Union[int, float], "list"),
+        (OmegaConf.structured(UnionClass), Union[int, str], "foo"),
+    ],
+)
+def test_pickle_union(obj: Any, ref_type: Any, key: str) -> None:
+    import collections
+
+    from omegaconf._utils import get_union_types
+
+    with tempfile.TemporaryFile() as fp:
+        pickle.dump(obj, fp)
+        fp.flush()
+        fp.seek(0)
+        c = pickle.load(fp)
+        node = c._get_node(key)
+        assert obj == c
+        if isinstance(node, (DictConfig, ListConfig)):
+            assert node._metadata.element_type == ref_type
+        else:
+            assert node._metadata.ref_type == ref_type
+            assert collections.Counter(node.element_types) == collections.Counter(
+                get_union_types(ref_type)
+            )
```

Comment:
> Also, _bring_type_to_front is implemented in case there's a value that can be converted in more than 1 element_type for example, 1 can be a str and an int therefore, if 1 is an int it should wrap first a IntegerNode instead of StringNode which would be incorrect.

Being type to front modifies the element_types so with counter it checks id the list has the same elements

Reply:
Just a note that this may be changed (see my comment regarding modifying element_types)

Reply:
Resolving since the counter is not needed anymore